File: src/main/java/net/nicoulaj/idea/markdown/lang/parser/sequentialparsers/SequentialParser.java
/*
 * Copyright (c) 2011-2014 Julien Nicoulaud <julien.nicoulaud@gmail.com>
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package net.nicoulaj.idea.markdown.lang.parser.sequentialparsers;

import com.intellij.openapi.util.TextRange;
import com.intellij.util.containers.ContainerUtil;
import net.nicoulaj.idea.markdown.lang.IElementType;
import net.nicoulaj.idea.markdown.lang.parser.TokensCache;
import org.jetbrains.annotations.NotNull;

import java.util.Collection;

public interface SequentialParser {

    ParsingResult parse(@NotNull TokensCache tokens, @NotNull Collection<TextRange> rangesToGlue);

    static class ParsingResult {
        @NotNull
        public final Collection<Node> parsedNodes;
        @NotNull
        public final Collection<Collection<TextRange>> rangesToProcessFurther;

        public ParsingResult() {
            this(ContainerUtil.<Node>newArrayList(), ContainerUtil.<Collection<TextRange>>newArrayList());
        }

        private ParsingResult(
                @NotNull Collection<Node> parsedNodes,
                @NotNull
                Collection<Collection<TextRange>> rangesToProcessFurther) {
            this.parsedNodes = parsedNodes;
            this.rangesToProcessFurther = rangesToProcessFurther;
        }

        public ParsingResult withNode(@NotNull Node result) {
            parsedNodes.add(result);
            return this;
        }

        public ParsingResult withNodes(@NotNull Collection<Node> parsedNodes) {
            this.parsedNodes.addAll(parsedNodes);
            return this;
        }

        public ParsingResult withFurtherProcessing(@NotNull Collection<TextRange> ranges) {
            rangesToProcessFurther.add(ranges);
            return this;
        }

    }

    static class Node {
        @NotNull
        public final TextRange range;
        @NotNull
        public final IElementType type;

        public Node(@NotNull TextRange range, @NotNull IElementType type) {
            this.range = range;
            this.type = type;
        }
    }

}

File: src/main/java/net/nicoulaj/idea/markdown/lang/parser/sequentialparsers/SequentialParser.kt
/*
 * Copyright (c) 2011-2014 Julien Nicoulaud <julien.nicoulaud@gmail.com>
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package net.nicoulaj.idea.markdown.lang.parser.sequentialparsers

import net.nicoulaj.idea.markdown.lang.IElementType
import net.nicoulaj.idea.markdown.lang.parser.TokensCache
import java.util.ArrayList

public trait SequentialParser {

    public fun parse(tokens: TokensCache, rangesToGlue: Collection<Range<Int>>): ParsingResult

    class object {

        public class Node(public val range: Range<Int>, public val type: IElementType)

        public class ParsingResult {

            private val _parsedNodes : MutableCollection<Node> = ArrayList()
            public val parsedNodes : Collection<Node>
                get() = _parsedNodes

            private val _rangesToProcessFurther : MutableCollection<Collection<Range<Int>>> = ArrayList()
            public val rangesToProcessFurther : Collection<Collection<Range<Int>>>
                get() = _rangesToProcessFurther

            public fun withNode(result: Node): ParsingResult {
                _parsedNodes.add(result)
                return this
            }

            public fun withNodes(parsedNodes: Collection<Node>): ParsingResult {
                _parsedNodes.addAll(parsedNodes)
                return this
            }

            public fun withFurtherProcessing(ranges: Collection<Range<Int>>): ParsingResult {
                _rangesToProcessFurther.add(ranges)
                return this
            }

        }

    }

}
